### PR TITLE
Expose more low level Index operations

### PR DESCRIPTION
**Pull request by nulltoken**
Besides adding/removing index entries by paths, this also introduces a way to add to the `Index` a `Blob`. 

This should facilitate scenarios such as partial staging (#195) for instance.

An additional commit contains a test that demonstrates how to perform a `git add --all` and some code to fix the current implementation of `repo.Stage()`.

### CODE REVIEW

File: LibGit2Sharp.Tests/IndexFixture.cs [hunk 1]
```diff
@@ -373,5 +374,124 @@ public void CanClearTheIndex()
...
+        public void AddingAnEntryToTheIndexFromAUnknwonFileInTheWorkdirThrows()
+        {
+            var path = SandboxStandardTestRepoGitDir();
+            using (var repo = new Repository(path))
+            {
+                const string filePath = "i_dont_exist.txt";
+                var before = repo.RetrieveStatus(filePath);
+                Assert.Equal(FileStatus.Nonexistent, before);
+
+                Assert.Throws<LibGit2SharpException>(() => repo.Index.Add(filePath));
```
Comment (nulltoken) [inline LibGit2Sharp.Tests/IndexFixture.cs]: This will throw a `NotFoundException` once #842 is merged

File: LibGit2Sharp.Tests/IndexFixture.cs [hunk 2]
```diff
@@ -373,5 +374,124 @@ public void CanClearTheIndex()
...
+                {
+                    case ChangeKind.Unmodified:
+                        continue;
+
+                    case ChangeKind.Added:
+                    case ChangeKind.Modified:
+                        repo.Index.Add(treeEntryChanges.Path);
+                        continue;
+
+                    case ChangeKind.Deleted:
```
Comment (Therzok) [inline LibGit2Sharp.Tests/IndexFixture.cs]: Shouldn't ChangeKind.Missing also fit here?

Comment (nulltoken) [inline LibGit2Sharp.Tests/IndexFixture.cs]: This is a `ChangeKind` enumeration, not a `FileStatus` one :wink:

Comment (Therzok) [inline LibGit2Sharp.Tests/IndexFixture.cs]: Oh, yeah. :dancers:

Comment (nulltoken) [inline LibGit2Sharp.Tests/IndexFixture.cs]: @Therzok Actually, that was a good lead! That led me to think about type changes.

I've added two related corner cases. One of them put under a flaw in the logic. Removals from the index should happen before additions (the index won't allow both `1.txt` and `1.txt/sneaky` entries to co-exist)

File: LibGit2Sharp/Repository.cs [hunk 3]
```diff
@@ -1478,29 +1478,34 @@ public void Stage(IEnumerable<string> paths, StageOptions stageOptions)
...
+            {
+                RemoveFromIndex(treeEntryChanges.Path);
+            }
 
-                    default:
-                        throw new InvalidOperationException(
-                            string.Format(CultureInfo.InvariantCulture, "Entry '{0}' bears an unexpected ChangeKind '{1}'", treeEntryChanges.Path, treeEntryChanges.Status));
-                }
+            foreach (TreeEntryChanges treeEntryChanges in changes
+                .Where(tec => tec.Status == ChangeKind.Added || tec.Status == ChangeKind.Modified))
```
Comment (dahlbyk) [inline LibGit2Sharp/Repository.cs]: This seems like a R#-ish. If we already have an outer `foreach` loop, I tend to just use an `if` instead of calling out to `Where()`. A microoptimization, if nothing else.

Comment (nulltoken) [inline LibGit2Sharp/Repository.cs]: Better?

**Comment by nulltoken** [comment]
Good to merge?
